source/gems/dry-validation/comparison-with-activemodel.html.md

```diff
@@ -0,0 +1,606 @@
+
+### 2.12 validates_with
+
+TODO expand
+
+### 2.13 validates_each
+
+TODO expand
+
+### 3. Common Validation Options
+
+These are the common options allowed by ActiveModel validations, and their equivalents in dry-validation
+
+**3.1 `:allow_nil`**
+
+Simply use `maybe` instead of `required` when defining your rules.
+
+**ActiveModel Validation**
+
+```ruby
+validates :attr, length: { minimum: int, allow_nil: true }
+```
+
+**dry-validation**
+
+```ruby
+required(:attr).filled(int?, min_size?: int)
+```
+
+**3.2  `:allow_blank`**
+
+In dry-validation you will need to use a block when defining your rule instead of `filled`, and include the `.empty?` predicate into your rule.
+
+**ActiveModel Validation**
+
+```ruby
+validates :attr, length: { minimum: int, allow_blank: true }
+```
+
+**dry-validation**
+
+```ruby
+required(:attr) { empty? | int? & min_size?: int )
+```
+
+
+**3.3 `:message`**
+
+Custom messages are implemented through a separate YAMl file. See [Error Messages](http://dry-rb.org/gems/dry-validation/error-messages/) for full instructions.
+
+**3.4 `:on`**
+
+In dry-validation, validations are defined in schemas. You can create separate schemas for various states (e.g UserCreateSchema, UserUpdateSchema) and then choose the correct schema to run in the relevant action.
+
+You can keep your schema code nice and DRY by [reusing schemas](gems/dry-validation/reusing-schemas/).
+
+### 4. Conditional Validation
+In ActiveModel you can use `:if` or `:unless` to only perform a validation based on the result of a proc or method.
+
+A simple schema can look like this:
+
+```ruby
+validates :card_number, presence: true, if: :paid_with_card?
+
+def paid_with_card?
+  payment_type == "card"
+end
+```
+
+To achieve this in dry-validation you can use [high-level rules](/gems/dry-validation/high-level-rules/).
+
+.1. Declare a rule for each of the attributes you need to reference:
+
+```ruby
+required(:payment_type).filled(inclusion?: ["card", "cash", "cheque"])
+required(:card_number).maybe
+```
+
+.2. Declare a custom predicate to check if `payment_type == 'card'`:
+
+```ruby
+def paid_with_card?(value)
+  value == "card"
+end
+```
+
+.3. Declare a high level rule to require the card number if `payment_type == 'card'`:
+
+```ruby
+rule(require_card_number: [:card_number, :payment_type]) do |card_number, payment_type|
+  payment_type.paid_with_card? > card_number.filled?
+end
+```
+
+Put it all together and you get:
+```ruby
+Dry::Validation.Schema do
+  required(:payment_type).filled(inclusion?: ["card", "cash", "cheque"])
+  required(:card_number).maybe
+
+  rule(require_card_number: [:card_number, :payment_type]) do |card_number, payment_type|
+    payment_type.paid_with_card? > card_number.filled?
+  end
+
+  configure do
+    def paid_with_card?(value)
+      value == "card"
+    end
+  end
```

Comment:
Hah it's OK, we should simply decide what exactly we want to explain - how to do implication *via* a custom predicate or just the implication alone. If we want to just go with the latter then we need to tweak AM example to use a proc in `if` option (IIRC it's supported?)